File: src/main/java/me/windyteam/kura/utils/mc/ChatUtil.java
package me.windyteam.kura.utils.mc;

import me.windyteam.kura.Kura;
import me.windyteam.kura.gui.Notification;
import me.windyteam.kura.utils.Utils;
import me.windyteam.kura.utils.Wrapper;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.GuiNewChat;
import net.minecraft.client.gui.ScaledResolution;
import net.minecraft.launchwrapper.LogWrapper;
import net.minecraft.network.play.client.CPacketChatMessage;
import net.minecraft.util.text.TextComponentString;
import net.minecraftforge.fml.relauncher.Side;
import net.minecraftforge.fml.relauncher.SideOnly;

import java.util.ArrayList;

public class ChatUtil {

    public static final int DeleteID = 94423;
    public static ArrayList<Notification> notifications = new ArrayList<>();
    public static char SECTIONSIGN = '\u00A7';

    public static String BLACK = SECTIONSIGN + "0";
    public static String DARK_BLUE = SECTIONSIGN + "1";
    public static String DARK_GREEN = SECTIONSIGN + "2";
    public static String DARK_AQUA = SECTIONSIGN + "3";
    public static String DARK_RED = SECTIONSIGN + "4";
    public static String DARK_PURPLE = SECTIONSIGN + "5";
    public static String GOLD = SECTIONSIGN + "6";
    public static String GRAY = SECTIONSIGN + "7";
    public static String DARK_GRAY = SECTIONSIGN + "8";
    public static String BLUE = SECTIONSIGN + "9";
    public static String GREEN = SECTIONSIGN + "a";
    public static String AQUA = SECTIONSIGN + "b";
    public static String RED = SECTIONSIGN + "c";
    public static String LIGHT_PURPLE = SECTIONSIGN + "d";
    public static String YELLOW = SECTIONSIGN + "e";
    public static String WHITE = SECTIONSIGN + "f";
    public static String OBFUSCATED = SECTIONSIGN + "k";
    public static String BOLD = SECTIONSIGN + "l";
    public static String STRIKE_THROUGH = SECTIONSIGN + "m";
    public static String UNDER_LINE = SECTIONSIGN + "n";
    public static String ITALIC = SECTIONSIGN + "o";
    public static String RESET = SECTIONSIGN + "r";

    public static String colorMSG = SECTIONSIGN + "r";
    public static String colorKANJI = SECTIONSIGN + "b";
    public static String colorWarn = SECTIONSIGN + "6" + SECTIONSIGN + "l";
    public static String colorError = SECTIONSIGN + "4" + SECTIONSIGN + "l";
    public static String colorBracket = SECTIONSIGN + "7";

    public static void clear() {
        notifications.clear();
    }

    public static void sendClientMessage(String message, Notification.Type type) {
        if (notifications.size() > 8) {
            notifications.remove(0);
        }
        notifications.add(new Notification(message, type));
    }

    public static void drawNotifications() {
        try {
            ScaledResolution res = new ScaledResolution(Minecraft.getMinecraft());
            double startY = res.getScaledHeight() - 25;
            final double lastY = startY;
            for (int i = 0; i < notifications.size(); i++) {
                Notification not = notifications.get(i);
                if (not.shouldDelete()) {
                    notifications.remove(not);
                    for (int cao = 0; cao > not.width; cao--) {
                        not.animationX = cao - not.width;
                    }
                    startY += not.getHeight() + 3;
                }
                not.draw(startY, lastY);
                for (int cao = 0; cao < not.width; cao++) {
                    not.animationX = cao + not.width;
                }
                startY -= not.getHeight() + 2;
            }
        } catch (Throwable ignored) {

        }
    }

    public static String bracketBuilder(String kanji) {
        return RESET + colorBracket + "[" + RESET + kanji + colorBracket + "] " + RESET;
    }

    public static void printRawMessage(String message) {
        if (Utils.nullCheck()) return;
        ChatMessage(message);
    }

    public static void printMessage(String message) {
        printRawMessage(bracketBuilder(colorKANJI + Kura.KANJI) + RESET + colorMSG + message);
    }

    public static void printWarnMessage(String message) {
        printRawMessage(bracketBuilder(colorKANJI + Kura.KANJI) + bracketBuilder(colorWarn + "WARN") + RESET + colorMSG + message);
    }

    public static void printErrorMessage(String message) {
        printRawMessage(bracketBuilder(colorKANJI + Kura.KANJI) + bracketBuilder(colorError + "ERROR") + RESET + colorMSG + message);
    }

    public static void sendRawMessage(String message) {
        printRawMessage(message);
    }

    public static void sendMessage(String message) {
        printMessage(message);
    }

    public static void sendDisablerDebugMessage(String message){
        sendMessage("[DISABLER] " + message);
    }

    public static void sendMessage(String[] message) {
        for (String msg : message) {
            if (msg != null) {
                sendMessage(msg);
            }
        }
    }

    public static void sendErrorMessage(String message) {
        printErrorMessage(message);
    }

    public static void sendWarnMessage(String message) {
        printWarnMessage(message);
    }

    public static void sendServerMessage(String message) {
        if (Minecraft.getMinecraft().player != null) {
            Wrapper.getPlayer().connection.sendPacket(new CPacketChatMessage(message));
        } else {
            LogWrapper.warning("Could not send server message: \"" + message + "\"");
        }
    }

    @SideOnly(Side.CLIENT)
    public static void sendSpamlessMessage(String message) {
        if (Utils.nullCheck()) return;
        final GuiNewChat chat = Wrapper.mc.ingameGUI.getChatGUI();
        chat.printChatMessageWithOptionalDeletion(new TextComponentString(message), DeleteID);
    }

    @SideOnly(Side.CLIENT)
    public static void sendSpamlessMessage(int messageID, String message) {
        if (Utils.nullCheck()) return;
        final GuiNewChat chat = Wrapper.mc.ingameGUI.getChatGUI();
        chat.printChatMessageWithOptionalDeletion(new TextComponentString(message), messageID);
    }

    public static void ChatMessage(String message) {
        Wrapper.mc.ingameGUI.getChatGUI().printChatMessage(new TextComponentString(message));
    }

    public static void sendNoSpamErrorMessage(String message) {
        sendNoSpamRawChatMessage(SECTIONSIGN + "7[" + SECTIONSIGN + "4" + SECTIONSIGN + "lERROR" + SECTIONSIGN + "7] " + SECTIONSIGN + "r" + message);
    }

    public static void sendNoSpamErrorMessage(String message, int messageID) {
        sendNoSpamRawChatMessage(SECTIONSIGN + "7[" + SECTIONSIGN + "4" + SECTIONSIGN + "lERROR" + SECTIONSIGN + "7] " + SECTIONSIGN + "r" + message, messageID);
    }

    public static void sendNoSpamRawChatMessage(String message) {
        sendSpamlessMessage(message);
    }

    public static void sendNoSpamRawChatMessage(String message, int messageID) {
        sendSpamlessMessage(messageID, message);
    }

    public static class NoSpam {
        public static void sendMessage(String message, int messageID) {
            sendRawChatMessage(bracketBuilder(colorKANJI + Kura.KANJI) + RESET + colorMSG + message, messageID);
        }

        public static void sendMessage(String message) {
            sendRawChatMessage(bracketBuilder(colorKANJI + Kura.KANJI) + RESET + colorMSG + message);
        }

        public static void sendMessage(String[] messages) {
            sendMessage("");
            for (String s : messages) sendRawChatMessage(s);
        }

        public static void sendErrorMessage(String message) {
            sendRawChatMessage(bracketBuilder(colorKANJI + Kura.KANJI) + bracketBuilder(colorError + "ERROR") + RESET + colorMSG + message);
        }

        public static void sendErrorMessage(String message, int messageID) {
            sendRawChatMessage(bracketBuilder(colorKANJI + Kura.KANJI) + bracketBuilder(colorError + "ERROR") + RESET + colorMSG + message, messageID);
        }

        public static void sendWarnMessage(String message) {
            sendRawChatMessage(bracketBuilder(colorKANJI + Kura.KANJI) + bracketBuilder(colorWarn + "WARN") + RESET + colorMSG + message);
        }

        public static void sendWarnMessage(String message, int messageID) {
            sendRawChatMessage(bracketBuilder(colorKANJI + Kura.KANJI) + bracketBuilder(colorWarn + "WARN") + RESET + colorMSG + message, messageID);
        }

        public static void sendRawChatMessage(String message) {
            sendSpamlessMessage(message);
        }

        public static void sendRawChatMessage(String message, int messageID) {
            sendSpamlessMessage(messageID, message);
        }
    }
}

File: src/main/java/me/windyteam/kura/utils/mc/ChatUtil.kt
package me.windyteam.kura.utils.mc

import com.mojang.realmsclient.gui.ChatFormatting
import me.windyteam.kura.Kura
import me.windyteam.kura.Kura.Companion.fontRenderer
import me.windyteam.kura.gui.Notification
import me.windyteam.kura.module.modules.client.Colors.chatColorMode
import me.windyteam.kura.utils.Utils
import me.windyteam.kura.utils.Wrapper
import net.minecraft.client.Minecraft
import net.minecraft.client.gui.ScaledResolution
import net.minecraft.launchwrapper.LogWrapper
import net.minecraft.network.play.client.CPacketChatMessage
import net.minecraft.util.text.TextComponentString
import net.minecraftforge.fml.relauncher.Side
import net.minecraftforge.fml.relauncher.SideOnly

object ChatUtil {
    const val DeleteID = 94423
    var notifications = ArrayList<Notification>()
    @JvmField
    var SECTIONSIGN = '§'
    var BLACK = SECTIONSIGN.toString() + "0"
    var DARK_BLUE = SECTIONSIGN.toString() + "1"
    var DARK_GREEN = SECTIONSIGN.toString() + "2"
    var DARK_AQUA = SECTIONSIGN.toString() + "3"
    var DARK_RED = SECTIONSIGN.toString() + "4"
    var DARK_PURPLE = SECTIONSIGN.toString() + "5"
    var GOLD = SECTIONSIGN.toString() + "6"
    var GRAY = SECTIONSIGN.toString() + "7"
    var DARK_GRAY = SECTIONSIGN.toString() + "8"
    var BLUE = SECTIONSIGN.toString() + "9"
    var GREEN = SECTIONSIGN.toString() + "a"
    var AQUA = SECTIONSIGN.toString() + "b"
    var RED = SECTIONSIGN.toString() + "c"
    var LIGHT_PURPLE = SECTIONSIGN.toString() + "d"
    var YELLOW = SECTIONSIGN.toString() + "e"
    var WHITE = SECTIONSIGN.toString() + "f"
    var OBFUSCATED = SECTIONSIGN.toString() + "k"
    var BOLD = SECTIONSIGN.toString() + "l"
    var STRIKE_THROUGH = SECTIONSIGN.toString() + "m"
    var UNDER_LINE = SECTIONSIGN.toString() + "n"
    var ITALIC = SECTIONSIGN.toString() + "o"
    var RESET = SECTIONSIGN.toString() + "r"
    var colorMSG = SECTIONSIGN.toString() + "r"
    var colorKANJI = SECTIONSIGN.toString() + "b"
    var colorWarn = SECTIONSIGN.toString() + "6" + SECTIONSIGN + "l"
    var colorError = SECTIONSIGN.toString() + "4" + SECTIONSIGN + "l"
    var colorBracket = SECTIONSIGN.toString() + "7"
    fun clear() {
        notifications.clear()
    }

    @JvmStatic
    fun sendClientMessage(message: String?, type: Notification.Type?) {
        if (notifications.size > 8) {
            notifications.removeAt(0)
        }
        notifications.add(Notification(message, type))
    }

    fun drawNotifications() {
        try {
            val res = ScaledResolution(Minecraft.getMinecraft())
            var startY = res.scaledHeight - 25.0000
            val lastY = startY
            for (i in notifications.indices) {
                val not = notifications[i]
                if (not.shouldDelete()) {
                    notifications.remove(not)
                    var cao = 0.0000
                    while (cao > not.width) {
                        not.animationX = cao - not.width
                        cao--
                    }
                    startY += not.getHeight() + 3.0000
                }
                not.draw(startY, lastY)
                var cao = 0.0000
                while (cao < not.width) {
                    not.animationX = cao + not.width
                    cao++
                }
                startY -= not.getHeight() + 2.0000
            }
        } catch (ignored: Throwable) {
        }
    }

    @JvmStatic
    fun getStringWidth(text: String): Int {
        return fontRenderer!!.getStringWidth(
            text
        )
    }

    fun bracketBuilder(kanji: String): String {
        return RESET + colorBracket + "[" + RESET + kanji + colorBracket + "] " + RESET
    }

    fun printRawMessage(message: String?) {
        if (Utils.nullCheck()) return
        ChatMessage(message)
    }

    fun printMessage(message: String) {
        printRawMessage("[" + chatColorMode() + Kura.KANJI + ChatFormatting.RESET + "] " + RESET + colorMSG + message)
    }

    fun printWarnMessage(message: String) {
        printRawMessage(bracketBuilder(colorKANJI + Kura.KANJI) + bracketBuilder(colorWarn + "WARN") + RESET + colorMSG + message)
    }

    fun printErrorMessage(message: String) {
        printRawMessage(bracketBuilder(colorKANJI + Kura.KANJI) + bracketBuilder(colorError + "ERROR") + RESET + colorMSG + message)
    }

    fun sendRawMessage(message: String?) {
        printRawMessage(message)
    }

    @JvmStatic
    fun sendMessage(message: String) {
        printMessage(message)
    }

    @JvmStatic
    fun sendDisablerDebugMessage(message: String) {
        sendMessage("[DISABLER] $message")
    }

    @JvmStatic
    fun sendMessage(message: Array<String?>) {
        for (msg in message) {
            if (msg != null) {
                sendMessage(msg)
            }
        }
    }

    @JvmStatic
    fun sendErrorMessage(message: String) {
        printErrorMessage(message)
    }

    fun sendWarnMessage(message: String) {
        printWarnMessage(message)
    }

    @JvmStatic
    fun sendServerMessage(message: String) {
        if (Minecraft.getMinecraft().player != null) {
            Wrapper.getPlayer().connection.sendPacket(CPacketChatMessage(message))
        } else {
            LogWrapper.warning("Could not send server message: \"$message\"")
        }
    }

    @SideOnly(Side.CLIENT)
    fun sendSpamlessMessage(message: String?) {
        if (Utils.nullCheck()) return
        val chat = Wrapper.mc.ingameGUI.chatGUI
        chat.printChatMessageWithOptionalDeletion(TextComponentString(message), DeleteID)
    }

    @SideOnly(Side.CLIENT)
    fun sendSpamlessMessage(messageID: Int, message: String?) {
        if (Utils.nullCheck()) return
        val chat = Wrapper.mc.ingameGUI.chatGUI
        chat.printChatMessageWithOptionalDeletion(TextComponentString(message), messageID)
    }

    fun ChatMessage(message: String?) {
        Wrapper.mc.ingameGUI.chatGUI.printChatMessage(TextComponentString(message))
    }

    fun sendNoSpamErrorMessage(message: String) {
        sendNoSpamRawChatMessage(SECTIONSIGN.toString() + "7[" + SECTIONSIGN + "4" + SECTIONSIGN + "lERROR" + SECTIONSIGN + "7] " + SECTIONSIGN + "r" + message)
    }

    fun sendNoSpamErrorMessage(message: String, messageID: Int) {
        sendNoSpamRawChatMessage(
            SECTIONSIGN.toString() + "7[" + SECTIONSIGN + "4" + SECTIONSIGN + "lERROR" + SECTIONSIGN + "7] " + SECTIONSIGN + "r" + message,
            messageID
        )
    }

    fun sendNoSpamRawChatMessage(message: String?) {
        sendSpamlessMessage(message)
    }

    fun sendNoSpamRawChatMessage(message: String?, messageID: Int) {
        sendSpamlessMessage(messageID, message)
    }

    object NoSpam {
        fun sendMessage(message: String, messageID: Int) {
            sendRawChatMessage(bracketBuilder(colorKANJI + Kura.KANJI) + RESET + colorMSG + message, messageID)
        }

        @JvmStatic
        fun sendMessage(message: String) {
            sendRawChatMessage("[" + chatColorMode() + Kura.KANJI + ChatFormatting.RESET + "] " + RESET + colorMSG + message)
        }

        fun sendMessage(messages: Array<String?>) {
            sendMessage("")
            for (s in messages) sendRawChatMessage(s)
        }

        @JvmStatic
        fun sendErrorMessage(message: String) {
            sendRawChatMessage(bracketBuilder(colorKANJI + Kura.KANJI) + bracketBuilder(colorError + "ERROR") + RESET + colorMSG + message)
        }

        fun sendErrorMessage(message: String, messageID: Int) {
            sendRawChatMessage(
                bracketBuilder(colorKANJI + Kura.KANJI) + bracketBuilder(colorError + "ERROR") + RESET + colorMSG + message,
                messageID
            )
        }

        @JvmStatic
        fun sendWarnMessage(message: String) {
            sendRawChatMessage(bracketBuilder(colorKANJI + Kura.KANJI) + bracketBuilder(colorWarn + "WARN") + RESET + colorMSG + message)
        }

        fun sendWarnMessage(message: String, messageID: Int) {
            sendRawChatMessage(
                bracketBuilder(colorKANJI + Kura.KANJI) + bracketBuilder(colorWarn + "WARN") + RESET + colorMSG + message,
                messageID
            )
        }

        fun sendRawChatMessage(message: String?) {
            sendSpamlessMessage(message)
        }

        fun sendRawChatMessage(message: String?, messageID: Int) {
            sendSpamlessMessage(messageID, message)
        }
    }
}
